Add path enumerable trait

bennothommo opened the pull request:
Implements path enumeration hierarchy for models, operating similar to nested set models but not requiring the rigid left/right values.

This is in preparation for some work towards database storage of media items.

bennothommo committed:
Add path enumerable trait

Implements path enumeration hierarchy for models, operating similar to nested set models but not requiring the rigid left/right values.

bennothommo committed:
Add getNested method and test

bennothommo committed:
Fix a couple of hard references to path column

LukeTowers commented on src/Database/Traits/PathEnumerable.php:
```diff
@@ -0,0 +1,322 @@
+            return $this->parent->{$this->getPathColumnName()} . '/' . $this->id;
+        }
+
+        return '/' . $this->id;
```
@bennothommo Does this have to be hardcoded to reference the `id` column? What if I wanted to use a custom column name, say `folder` *cough cough*

bennothommo replied:
We can make a constant for the heathens.

LukeTowers replied:
What are the practical reasons for not wanting to use the actual folder names for the Media Library use case?

bennothommo replied:
@LukeTowers probably nothing in terms of queries and functionality, but it could potentially result in larger path strings which would mean less levels of directories before you hit MySQL's string limits. At least with IDs you'll have a lot more levels before hitting those limits.

LukeTowers replied:
Fair enough, do you think it would be worth adding in a constant for the max length of the path to surface that before it causes issues?

bennothommo replied:
@LukeTowers I would say if the intention is to try and catch it before it becomes a problem, it's probably worth querying the schema directly and working out the maximum length, as opposed to an arbitrary max length constant. But either way, it will mean that the "segment" column (id/folder/etc.) would need to be configurable in some fashion, so yeah, it's either a constant or a method to get that column.

bennothommo replied:
@LukeTowers I've added the ability to define a custom segment column for the path.

bennothommo replied:
Further to this, I opted not to try and find the maximum length of the path column in the schema as the only way to do so is to run an SQL query, which means it would call this every page load that a enumerable path model is used (unless it were cached).

I think this is something we can opt to document as opposed to trying to technically handle - most people, I imagine, would opt to use the ID paths, and for people who know how to change the segment column, they (should?) be knowledgeable enough to ensure their DB schema accounts for the potentially larger path values.

LukeTowers commented:
@bennothommo 👀

bennothommo committed:
Merge branch 'develop' into wip/path-enumerable-trait

bennothommo committed:
Add ability to define a custom segment column.

This allows the path to be made up of segments from a different field, such as folder, or name. By default, path segments are derived from the primary key (usually an ID).

LukeTowers commented:
@bennothommo let me know when you want me to review this!

bennothommo commented:
@LukeTowers it's good to go :)

bennothommo committed:
Prevent forward-slashes in segment column from breaking hierarachy

bennothommo committed:
Fix comments

LukeTowers reviewed (approved):
LGTM @bennothommo, targeting next release pending docs PR for this feature :)

bennothommo committed:
Merge remote-tracking branch 'origin/develop' into wip/path-enumerable-trait

bennothommo commented:
@LukeTowers not yet - I'm toying with the idea of unifying the tree traits in some fashion so they share a common base level of functionality and API.

LukeTowers commented:
@bennothommo is that something that needs to happen now or could we move forward with merging this so that we can get metadata support added to the Media Library?

bennothommo commented:
@LukeTowers my concern is that we'll have difficulty changing the API later.

LukeTowers commented:
@bennothommo status on this?

bennothommo commented:
Just waiting for docs. I mentioned previously trying to attempt to unify the tree APIs, but I probably won't do it for this branch. It might be worth looking into next major release.

LukeTowers commented:
@bennothommo sounds good to me, lets get the docs done up so we can roll this into the next release then. Would be nice to make some progress on the media library metadata.

LukeTowers commented:
@bennothommo ping

bennothommo committed:
Merge branch 'develop' into wip/path-enumerable-trait

LukeTowers commented:
@bennothommo what specifically needs to be documented?

bennothommo commented:
Docs PR is here:  https://github.com/wintercms/docs/pull/227
